### DISCUSSION TITLE
Lemma_ empty after training with spaCy 3.0rc2 and en_core_web_trf

### DISCUSSION THREAD

**Discussion by AlexTo** [explosion/spaCy#6585] [Help: Coding & Implementations]
## Info about spaCy

- **spaCy version:** 3.0.0rc2
- **Platform:** Linux-5.4.0-58-generic-x86_64-with-debian-bullseye-sid
- **Python version:** 3.7.9
- **Pipelines:** en_core_web_trf (3.0.0a0)

I wanted to retrain the en_core_web_trf with my own dataset  and configuration attached below. The training datasets contain NER that I want to train with. I only want to improve the NER process.

The command that I used:

```
python -m spacy train ./config_0.0.2.cfg --paths.train ./spacy_train.spacy --paths.dev ./spacy_val.spacy --gpu-id=0 --output='models'
```

After training, I used the new model to create a nlp doc for e.g with the following content

```
I want to increase limit on my card # 5361238925377564 for certain duration of time. is it possible?
```
But the lemma_ are all empty strings

![image](https://user-images.githubusercontent.com/296212/102441051-9def7780-4075-11eb-98f8-0d31e8b0fb7f.png)

It works ok with model `en_core_web_trf `. Does anyone know what have I done wrong ?

Thank you 

[spacy_train.zip](https://github.com/explosion/spaCy/files/5706823/spacy_train.zip)
[spacy_val.zip](https://github.com/explosion/spaCy/files/5706835/spacy_val.zip)
[config_0.0.2.zip](https://github.com/explosion/spaCy/files/5706824/config_0.0.2.zip)

**Comment by ines** [comment]
Copy-pasting your config here so others don't have to download a zip:

```ini
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "en"
pipeline = ["transformer","ner"]
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null


[components]

[components.ner]
source = "en_core_web_trf"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v1"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = false
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}

[components.transformer]
source = "en_core_web_trf"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = "roberta-base"

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96

[components.transformer.model.tokenizer_config]
use_fast = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 500
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
ents_per_type = null
ents_f = 2.0
ents_p = 2.0
ents_r = 6.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null

[initialize.components]

[initialize.tokenizer]
```

From looking at your config, it seems like you're only including the `transformer` and `ner` components, and no `tagger`, `atrribute_ruler` and `lemmatizer`. So your model doesn't end up with a lemmatizer that assigns lemmas, and no annotations that can be used to determine the lemmas. See here for the components available in the `en_core_web_trf` pipeline: https://nightly.spacy.io/models/en

**Comment by AlexTo** [comment]
Hi Ines
Thanks for your quick reply, after your suggestion, I added the following to the above config

```
[nlp]
lang = "en"
pipeline = ["transformer", "tagger", "parser", "ner", "attribute_ruler", "lemmatizer"]

[components.tagger]
source = "en_core_web_trf"

[components.parser]
source = "en_core_web_trf"

[components.attribute_ruler]
source = "en_core_web_trf"

[components.lemmatizer]
source = "en_core_web_trf"

[training]
frozen_components = ["tagger", "parser", "attribute_ruler", "lemmatizer"]
```
Now during training, I could see more "columns" as follows:

![image](https://user-images.githubusercontent.com/296212/102444181-a9926c80-407c-11eb-9b86-d98b264369a2.png)

But the `LEMMA_ACC` and `DEP_LAS` are all 0. Should I be concerned about that ??

Thanks a lot

**Reply by adrianeboyd** [reply]
If your training data does not contain labeled dependency parses or lemmas, there's no way to evaluate on them, so you'll see 0.0 or meaningless scores (the unlabeled dependency evaluation is probably just showing that some of the roots are correct).

From the tagger scores, it looks like updating the shared tok2vec component just for NER is degrading the performance for the tagger, so it's probably similar for the parser.

You could automatically annotate some of your eval data with `en_core_web_trf` and compare the parser and lemmatizer evaluation before and after your NER training to see how large the difference is. It won't be a perfect comparison (not gold evaluation data), but if the numbers go down a lot, it's something to look into for your task.

In general, it's better to update the transformer pipelines with training data for all the components at the same time, so it might actually work better to automatically annotate the missing dependency parses for your data and then not freeze the tagger and parser components while training.

Training from silver data has its own drawbacks, so I'm not totally sure what the difference would look like, but it could be worth trying out.

**Reply by AlexTo** [reply]
Thanks a lot, I am trying to add the full pipeline and retraining the model. I usually work directly with deepnet, first time trying out spaCy ;)
